scripts/dotnet-cli-build/PublishTargets.cs

```diff
@@ -48,12 +52,111 @@ public static BuildTargetResult Publish(BuildTargetContext c)
             return c.Success();
         }
 
+        [Target]
+        public static BuildTargetResult FinalizeBuild(BuildTargetContext c)
+        {
+            if (CheckIfAllBuildsHavePublished())
+            {
+                string targetContainer = $"{Channel}/Binaries/Latest/";
+                string targetVersionFile = $"{targetContainer}{CliNuGetVersion}";
+                string leaseId = AzurePublisherTool.AcquireLeaseOnContainer();
+
+                // Prevent race conditions by dropping a version hint of what version this is. If we see this file
+                // and it is the same as our version then we know that a race happened where two+ builds finished 
+                // at the same time and someone already took care of publishing and we have no work to do.
+                if (AzurePublisherTool.IsLatestSpecifiedVersion(targetVersionFile))
+                {
+                    AzurePublisherTool.ReleaseLeaseOnContainer(leaseId);
```

Comment:
Close, the flow is as such (changed slightly between version of my PR):

1) try acquire a lease on the ```latest``` folder
2) Write version and do work
3) Release lease

If there is contention, the second build will wait and then see the version file once the lease expires and exit quickly. The only coordination object is the latest folder itself